Code review n.3 for Enrico Cestaro (by Matteo Carnelos)

matteocarnelos opened the pull request:
## Premise
**(It might seems a lot of lines but a portion has to be excluded, see "Files Submitted")**

For this review I've submitted code that I'm particularly proud of, that is the realization of the user experience, and in particular the responsiveness of the UI and the retention of data. Some changes have been made accordingly to suggestions coming from the [Moodle Forum](https://elearning.dei.unipd.it/mod/forum/discuss.php?d=38084).
## Files submitted
### 1. MainActivity
As you may already know is the starting point of the entire application. I've included it because it is particularly important for the initialization of storage modules and the displaying of first-launch helpers (I used helpers instead of pre-activities in order to make the interaction more fluid and fast). In this class **`isNotificationListenerEnabled(...)` and `openNotificationListenSettings()` don't need to be reviewed** because they are not made by me.
### 2. CreatePollActivity
This is the activity for the creation of new polls, it contains some custom animations and touches handling as well as the preparation and serialization of data to send to the main activity.
### 3. BinaryPoll
This class represents and manage poll objects, including the storing of the incremental poll number in the shared preferences of the app. The last mentioned is the most important function of the whole class and so the **"Getters" and "Setters" sections don't need to be reviewed**.
### 4. DataProvider
This class was submitted also in a previous review but since then I've made some important changes that made it revisable again. Improvements are:
1. The management of polls in a synchronized way, this because every poll object can be seen as a shared resource between all the provider users and so a lock must be acquired before accessing it.
2. The conversion and the storing of polls in the internal memory, every poll is in fact converted into JSON format before saving it to the internal memory of the devices and vice versa, during every app boot (the loading of saved polls is triggered in the MainActivity).

As for the last submitted file **I don't particularly need a review for what was already reviewed and for "Getters" and "Setters" sections**.

---
I hope I was clear, for any additional information see classes java docs or ask me here.
Good review😁

matteocarnelos committed:
Added files to review

EnricoCestaro commented on easypoll/src/main/java/com/eis0/easypoll/poll/BinaryPoll.java:
```diff
@@ -0,0 +1,300 @@
+     */
+    public static void loadPollsCountFromInternal() {
+        pollsCount = mSharedPreferences.getLong(POLLS_COUNT_KEY, 0);
+    }
```
I'm not sure about the reason why you are enabling mSharedPreferences to be null.
In the savePollsCountToInternal() if the value is null you simply don't do anything, not even an advertisement for the programmer, incorrect use of the methods could cause to not see the problem before the use of the loadPollsCountFromInternal(), when I assume that if the mSharedPreferences is null a NullPointerException will be thrown.

matteocarnelos replied:
> I'm not sure about the reason why you are enabling mSharedPreferences to be null.

I did that because the shared preferences pointer is obtained from the context of the application and this class doesn't have any reference to that, so two was the possible ways:
1. **Asking the context (or, better, directly the shared preferences pointer) in the constructor of the class.**
This solution is usually adopted when the resource needed is indispensable for the correct working of the class or it is needed during the creation of the object. However, in my case the storing feature is not indispensable, as the class will still work without it (obviously losing the storing feature if the user doesn't apply any other type of data retention), and it is not needed during creation. So I went for solution 2.
2. **Initially setting the shared preferences pointer null and letting the user set its value if needed.**
As I said before, the storing feature is meant to be optional, so this solution was the right one for my case.

> In the savePollsCountToInternal() if the value is null you simply don't do anything, not even an advertisement for the programmer

I chose to don't do anything because launching an exception was too much in my opinion, I usually launch exceptions when the working of the program is compromised and it can't go on, and this wasn't the case.
You're totally right about the advertisement for the programmer, I missed that. It could be done by informing the user in the specifics of the method that, if `null`, the method will terminate quietly. That's what I'll do, thanks!

> I assume that if the mSharedPreferences is null a NullPointerException will be thrown.

That's my fault, I missed the nullity check like I do in `savePollsCountToInternal()`. Will do!

EnricoCestaro replied:
Exceptional explanation, I had the impression that there was a reason for your choice considering the rest of the code; anyway I think that's all for this code review.

I really like your style, how you split the different methods (Getters, Setters etc...), the names of the variables you use, maybe you could add some more explicative descriptions for methods and classes, but I probably say it because of my attitude to writing poems for the simplest commands. It surely was easier for me understanding what you did considering that I already knew the purpose of those classes, but the order in the code helped a lot.

The most interesting part of the code is, in my opinion, the one which doesn't operate directly with the UI, because even if it's made perfectly it's like a showcase of everything you can do in Android to create an Interface; still really helpful considering that right now I'm totally unable to do something similar.

matteocarnelos replied:
I really appreciate that, thanks a lot!😁

EnricoCestaro commented on easypoll/src/main/java/com/eis0/easypoll/poll/BinaryPoll.java:
```diff
@@ -0,0 +1,300 @@
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
```
I used an older version of your equals() method, just for curiosity why didn't you use the instanceof operator? Is it just because the result is the same, or you found this as a more proper way to do it?

matteocarnelos replied:
The `getClass()` way is more strict. With `instanceof` you will get `true` if one of the two classes is a subclass of the other, while with `getClass()` you will check that the two types are identical.
In this particular case using `instanceof` won't have caused any difference, but I usually prefer to be more strict on equals methods.

EnricoCestaro replied:
I was hoping in this kind of answer. Thanks

EnricoCestaro reviewed (commented):
First comments to the code (which is really well made)

EnricoCestaro commented on easypoll/src/main/java/com/eis0/easypoll/poll/BinaryPoll.java:
```diff
@@ -0,0 +1,300 @@
+
+    /**
+     * Set the SharedPreferences reference to which save the {@link #pollsCount} value.
+     *
```
A missing @param.
And speaking about more or less all the code I can see that you make good use of the @NonNull but only in some occasions, to be more coherent you should use it everywhere or change it with @Nullable where is the case

matteocarnelos replied:
That's a good point, I missed **@param** here and some annotations along the code. I'll add them as soon as I can. I thought that annotations in overridden methods would be inherited automatically but after some searches I found out that it is not, they need to be repeated in every sub-method. Thanks!

EnricoCestaro replied:
Perfect

matteocarnelos committed:
Applied suggested changes:

+ Added annotations
~ Modified specifics to be more informative
+ Added nullity checks

matteocarnelos committed:
Applied suggested changes:

+ Added @param tag
